Assignment_4.py

```diff
@@ -0,0 +1,318 @@
+#!/usr/bin/env python3
+import os
+import sys
+
+import numpy as np
+import pandas as pd
+import statsmodels.api as sm
+from plotly import express as px
+from plotly import figure_factory as ff
+from plotly import graph_objects as go
+from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
+from sklearn.metrics import confusion_matrix
+
+
+# Check if Response is Categorical or Continuous
+def Check_response_var(inp_df):
+    if inp_df.target.nunique() > 2:
+        return "Continuous"
+    else:
+        return "Boolean"
+
+
+# Check if Predictor is Categorical
+# Categorical have few unique values
+def is_pred_categorical(pred):
+    if pred.dtypes == "object" or pred.nunique() / pred.count() < 0.05:
+        return True
+    else:
+        return False
+
+
+# Generate Heatmap Categorical Predictor by Categorical Response
+def generate_heatmap(df, col, filename):
+    conf_matrix = confusion_matrix(df[col], df["target"])
+    fig_no_relationship = go.Figure(
+        data=go.Heatmap(z=conf_matrix, zmin=0, zmax=conf_matrix.max())
+    )
+    fig_no_relationship.update_layout(
+        title="Categorical Predictor by Categorical Response ",
+        xaxis_title="Response",
+        yaxis_title="Predictor",
+    )
+    fig_no_relationship.show()
+    fig_no_relationship.write_html(
+        file=filename,
+        include_plotlyjs="cdn",
+    )
+
+
+# Generate Violin plot when Continuous Response by Categorical Predictor
+def generate_violin_plot(df, col, filename):
+    group_labels = pd.unique(df.target)
+    fig_1 = go.Figure()
+    M = df[df["target"] == group_labels[0]][col]
+    B = df[df["target"] == group_labels[1]][col]
+    for curr_hist, curr_group in zip([M, B], group_labels):
+        fig_1.add_trace(
+            go.Violin(
+                x=np.repeat(curr_group, len(df)),
+                y=curr_hist,
+                name=int(curr_group),
+                box_visible=True,
+                meanline_visible=True,
+            )
+        )
+    fig_1.update_layout(
+        title="Continuous Response by Categorical Predictor",
+        xaxis_title="Groupings",
+        yaxis_title="Response",
+    )
+    # fig_1.show()
+    fig_1.write_html(
+        file=filename,
+        include_plotlyjs="cdn",
+    )
+
+
+def generate_dist_plot(df, col, filename):
+    group_labels = ["0", "1"]
+    # Create distribution plot with custom bin_size
+    M = df[df["target"] == 0][col]
+    B = df[df["target"] == 1][col]
+    fig_1 = ff.create_distplot([M, B], group_labels, bin_size=0.2)
+    fig_1.update_layout(
+        title="Continuous Predictor by Categorical Response",
+        xaxis_title="Predictor",
+        yaxis_title="Distribution",
+    )
+    # fig_1.show()
+    fig_1.write_html(
```

Comment:
The HTML page actually looks really nice, and it is quite easy to access everything from there. It is like having all the information related to the dataset on the same page.